File: backend/app/models/agent_chat.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


AgentAction = Literal[
    "general",
    "analyze_pick",
    "review_slip",
    "risk_check",
    "line_movement",
    "plain_english",
    "review_board",
]

AgentStatus = Literal[
    "ok",
    "line_moved",
    "not_enough_market_data",
    "injury_context_missing",
    "insufficient_context",
    "error",
]

AgentDecision = Literal["over", "under", "avoid"]
AgentRecommendation = Literal["keep", "recheck", "remove"]
LineupStatus = Literal["projected", "partial", "unavailable"]
LineupConfidence = Literal["high", "medium", "low"]
AgentVerdictEvidenceTone = Literal["positive", "neutral", "caution", "muted"]
AgentVerdictBreakdownTone = Literal["support", "caution", "neutral", "unavailable"]
AgentVerdictBreakdownSectionKey = Literal[
    "historical",
    "trend_role",
    "shooting",
    "variance",
    "schedule",
    "own_team_injuries",
    "lineup",
    "market",
    "projection",
]


class AgentPageContext(BaseModel):
    route: str = Field(default="/", description="Current frontend route")
    date: str | None = Field(default=None, description="Current board date")
    selected_teams: list[str] = Field(
        default_factory=list,
        description="Currently selected teams on the board",
    )


class AgentPickContext(BaseModel):
    player_name: str = Field(..., description="Player name")
    player_team: str = Field(default="", description="Player team")
    event_id: str = Field(..., description="Event identifier")
    home_team: str = Field(..., description="Home team")
    away_team: str = Field(..., description="Away team")
    commence_time: str = Field(..., description="Game start time")
    metric: str = Field(..., description="Prop metric")
    threshold: float = Field(..., description="Market threshold")
    direction: str = Field(..., description="Current pick direction")
    probability: float = Field(..., description="Current historical probability")
    n_games: int = Field(..., description="Sample size")
    projected_value: float | None = Field(default=None, description="Projection value")
    projected_minutes: float | None = Field(default=None, description="Projected minutes")
    edge: float | None = Field(default=None, description="Projection edge")


class AgentContext(BaseModel):
    page: AgentPageContext | None = Field(default=None, description="Page context")
    selected_pick: AgentPickContext | None = Field(
        default=None,
        description="Selected pick context",
    )
    visible_picks: list[AgentPickContext] = Field(
        default_factory=list,
        description="Visible picks on the current board",
    )
    bet_slip: list[AgentPickContext] = Field(
        default_factory=list,
        description="Current bet slip contents",
    )


class AgentQuickAction(BaseModel):
    action: AgentAction = Field(..., description="Quick action key")
    label: str = Field(..., description="Button label")
    prompt: str = Field(..., description="Prompt to submit for the action")


class AgentLineupTeamContext(BaseModel):
    team: str = Field(..., description="Team code")
    status: LineupStatus = Field(..., description="Current lineup state")
    confidence: LineupConfidence | None = Field(default=None, description="Consensus confidence")
    source_disagreement: bool = Field(default=False, description="Whether sources disagree")
    updated_at: str | None = Field(default=None, description="Latest lineup update timestamp")
    player_is_projected_starter: bool | None = Field(
        default=None,
        description="Whether the selected player is in the projected starters",
    )
    starters: list[str] = Field(default_factory=list, description="Projected starters")


class AgentLineupContext(BaseModel):
    summary: str = Field(default="", description="Human-readable lineup summary")
    freshness_risk: bool = Field(default=False, description="Whether lineup data is stale")
    player_team: AgentLineupTeamContext | None = Field(
        default=None,
        description="Selected player's team lineup context",
    )
    opponent_team: AgentLineupTeamContext | None = Field(
        default=None,
        description="Opponent lineup context",
    )


class AgentVerdictEvidenceStat(BaseModel):
    label: str = Field(..., description="Short stat label")
    value: str = Field(..., description="Formatted stat value")
    tone: AgentVerdictEvidenceTone | None = Field(
        default=None,
        description="positive | neutral | caution | muted",
    )


class AgentVerdictBreakdownSection(BaseModel):
    key: AgentVerdictBreakdownSectionKey = Field(..., description="Breakdown section key")
    label: str = Field(..., description="Human-readable section label")
    tone: AgentVerdictBreakdownTone = Field(
        ...,
        description="support | caution | neutral | unavailable",
    )
    reliability: float | None = Field(
        default=None,
        ge=0,
        le=1,
        description="Section reliability score when available",
    )
    signal_note: str = Field(..., description="Primary user-facing signal note")
    risk_note: str | None = Field(default=None, description="Material caution for this section")
    stats: list[AgentVerdictEvidenceStat] = Field(
        default_factory=list,
        description="Compact evidence stats for the section",
    )


class AgentVerdictBreakdown(BaseModel):
    sections: list[AgentVerdictBreakdownSection] = Field(
        default_factory=list,
        description="Ordered breakdown sections for the verdict",
    )


class AgentVerdictCard(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    subject: str = Field(..., description="Human-readable subject label")
    decision: AgentDecision = Field(..., description="Agent decision")
    confidence: float = Field(..., ge=0, le=1, description="Confidence score")
    model_probability: float = Field(..., ge=0, le=1, description="Modeled probability")
    market_implied_probability: float | None = Field(
        default=None,
        ge=0,
        le=1,
        description="Market implied probability for the queried side",
    )
    expected_value_pct: float | None = Field(
        default=None,
        description="Expected value percentage",
    )
    market_pricing_mode: str = Field(
        default="unavailable",
        description="How the live market priced the queried pick",
    )
    queried_line: float | None = Field(
        default=None,
        description="Original line requested for pricing",
    )
    best_line: float | None = Field(
        default=None,
        description="Best currently available live line",
    )
    available_lines: list[float] = Field(
        default_factory=list,
        description="Distinct live lines currently available",
    )
    best_book: str | None = Field(
        default=None,
        description="Best bookmaker for the live line",
    )
    best_odds: int | None = Field(
        default=None,
        description="Best available American odds",
    )
    summary: str = Field(..., description="Verdict summary")
    breakdown: AgentVerdictBreakdown | None = Field(
        default=None,
        description="Structured reasoning breakdown",
    )
    reasons: list[str] = Field(default_factory=list, description="Primary reasons")
    risk_factors: list[str] = Field(default_factory=list, description="Risk factors")
    lineup_context: AgentLineupContext | None = Field(
        default=None,
        description="Projected lineup context for the matchup",
    )
    recommendation: AgentRecommendation | None = Field(
        default=None,
        description="Slip review recommendation",
    )


class AgentSlipReview(BaseModel):
    items: list[AgentVerdictCard] = Field(
        default_factory=list,
        description="Per-leg verdicts",
    )
    summary: str = Field(..., description="Slip-level summary")
    keep_count: int = Field(default=0, ge=0, description="Count of keep recommendations")
    recheck_count: int = Field(
        default=0,
        ge=0,
        description="Count of recheck recommendations",
    )
    remove_count: int = Field(
        default=0,
        ge=0,
        description="Count of remove recommendations",
    )


class AgentChatRequest(BaseModel):
    thread: str = Field(default="", description="Client-managed thread identifier")
    message: str = Field(default="", description="User message")
    action: AgentAction = Field(default="general", description="Requested action")
    context: AgentContext | None = Field(default=None, description="UI context payload")


class AgentChatResponse(BaseModel):
    thread: str = Field(default="", description="Client-managed thread identifier")
    action: AgentAction = Field(..., description="Resolved action")
    status: AgentStatus = Field(..., description="Response status")
    reply: str = Field(..., description="Assistant reply text")
    verdict: AgentVerdictCard | None = Field(
        default=None,
        description="Single-pick verdict card",
    )
    slip_review: AgentSlipReview | None = Field(
        default=None,
        description="Slip review payload",
    )
    quick_actions: list[AgentQuickAction] = Field(
        default_factory=list,
        description="Suggested follow-up actions",
    )

File: frontend/lib/agent-chat.ts
import { z } from "zod";

import type { BetSlipPick } from "@/contexts/BetSlipContext";
import type { DailyPick } from "@/lib/schemas";

export const agentActionSchema = z.enum([
  "general",
  "analyze_pick",
  "review_slip",
  "risk_check",
  "line_movement",
  "plain_english",
  "review_board",
]);

export type AgentAction = z.infer<typeof agentActionSchema>;

export const agentPageContextSchema = z.object({
  route: z.string().default("/"),
  date: z.string().nullable().optional(),
  selected_teams: z.array(z.string()).default([]),
});

export type AgentPageContext = z.infer<typeof agentPageContextSchema>;

export const agentPickContextSchema = z.object({
  player_name: z.string(),
  player_team: z.string().default(""),
  event_id: z.string(),
  home_team: z.string(),
  away_team: z.string(),
  commence_time: z.string(),
  metric: z.string(),
  threshold: z.number(),
  direction: z.string(),
  probability: z.number(),
  n_games: z.number(),
  projected_value: z.number().nullable().optional(),
  projected_minutes: z.number().nullable().optional(),
  edge: z.number().nullable().optional(),
});

export type AgentPickContext = z.infer<typeof agentPickContextSchema>;

export const agentRequestContextSchema = z.object({
  page: agentPageContextSchema.nullable().optional(),
  selected_pick: agentPickContextSchema.nullable().optional(),
  visible_picks: z.array(agentPickContextSchema).default([]),
  bet_slip: z.array(agentPickContextSchema).default([]),
});

export type AgentRequestContext = z.infer<typeof agentRequestContextSchema>;

export const agentQuickActionSchema = z.object({
  action: agentActionSchema,
  label: z.string(),
  prompt: z.string(),
});

export type AgentQuickAction = z.infer<typeof agentQuickActionSchema>;

export const agentLineupTeamContextSchema = z.object({
  team: z.string(),
  status: z.enum(["projected", "partial", "unavailable"]),
  confidence: z.enum(["high", "medium", "low"]).nullable().optional(),
  source_disagreement: z.boolean().default(false),
  updated_at: z.string().nullable().optional(),
  player_is_projected_starter: z.boolean().nullable().optional(),
  starters: z.array(z.string()).default([]),
});

export type AgentLineupTeamContext = z.infer<typeof agentLineupTeamContextSchema>;

export const agentLineupContextSchema = z.object({
  summary: z.string().default(""),
  freshness_risk: z.boolean().default(false),
  player_team: agentLineupTeamContextSchema.nullable().optional(),
  opponent_team: agentLineupTeamContextSchema.nullable().optional(),
});

export type AgentLineupContext = z.infer<typeof agentLineupContextSchema>;

export const agentVerdictEvidenceStatSchema = z.object({
  label: z.string(),
  value: z.string(),
  tone: z.enum(["positive", "neutral", "caution", "muted"]).nullable().optional(),
});

export type AgentVerdictEvidenceStat = z.infer<typeof agentVerdictEvidenceStatSchema>;

export const agentVerdictBreakdownSectionSchema = z.object({
  key: z.enum([
    "historical",
    "trend_role",
    "shooting",
    "variance",
    "schedule",
    "own_team_injuries",
    "lineup",
    "market",
    "projection",
  ]),
  label: z.string(),
  tone: z.enum(["support", "caution", "neutral", "unavailable"]),
  reliability: z.number().nullable().optional(),
  signal_note: z.string(),
  risk_note: z.string().nullable().optional(),
  stats: z.array(agentVerdictEvidenceStatSchema).default([]),
});

export type AgentVerdictBreakdownSection = z.infer<typeof agentVerdictBreakdownSectionSchema>;

export const agentVerdictBreakdownSchema = z.object({
  sections: z.array(agentVerdictBreakdownSectionSchema).default([]),
});

export type AgentVerdictBreakdown = z.infer<typeof agentVerdictBreakdownSchema>;

export const agentVerdictCardSchema = z.object({
  subject: z.string(),
  decision: z.enum(["over", "under", "avoid"]),
  confidence: z.number(),
  model_probability: z.number(),
  market_implied_probability: z.number().nullable().optional(),
  expected_value_pct: z.number().nullable().optional(),
  market_pricing_mode: z.string().default("unavailable"),
  queried_line: z.number().nullable().optional(),
  best_line: z.number().nullable().optional(),
  available_lines: z.array(z.number()).default([]),
  best_book: z.string().nullable().optional(),
  best_odds: z.number().nullable().optional(),
  summary: z.string(),
  breakdown: agentVerdictBreakdownSchema.nullable().optional(),
  reasons: z.array(z.string()).default([]),
  risk_factors: z.array(z.string()).default([]),
  lineup_context: agentLineupContextSchema.nullable().optional(),
  recommendation: z.enum(["keep", "recheck", "remove"]).nullable().optional(),
});

export type AgentVerdictCard = z.infer<typeof agentVerdictCardSchema>;

export const agentSlipReviewSchema = z.object({
  items: z.array(agentVerdictCardSchema).default([]),
  summary: z.string(),
  keep_count: z.number(),
  recheck_count: z.number(),
  remove_count: z.number(),
});

export type AgentSlipReview = z.infer<typeof agentSlipReviewSchema>;

export const agentChatResponseSchema = z.object({
  thread: z.string(),
  action: agentActionSchema,
  status: z.enum([
    "ok",
    "line_moved",
    "not_enough_market_data",
    "injury_context_missing",
    "insufficient_context",
    "error",
  ]),
  reply: z.string(),
  verdict: agentVerdictCardSchema.nullable().optional(),
  slip_review: agentSlipReviewSchema.nullable().optional(),
  quick_actions: z.array(agentQuickActionSchema).default([]),
});

export type AgentChatResponse = z.infer<typeof agentChatResponseSchema>;

export interface AgentChatRequest {
  thread: string;
  message: string;
  action: AgentAction;
  context?: AgentRequestContext;
}

export interface AgentThreadMessage {
  id: string;
  role: "user" | "assistant";
  text: string;
  action?: AgentAction;
  status?: AgentChatResponse["status"];
  quick_actions?: AgentQuickAction[];
  verdict?: AgentVerdictCard | null;
  slip_review?: AgentSlipReview | null;
}

export const createAgentPickContextFromDailyPick = (
  pick: DailyPick,
): AgentPickContext => ({
  player_name: pick.player_name,
  player_team: pick.player_team || "",
  event_id: pick.event_id,
  home_team: pick.home_team,
  away_team: pick.away_team,
  commence_time: pick.commence_time,
  metric: pick.metric,
  threshold: pick.threshold,
  direction: pick.direction,
  probability: pick.probability,
  n_games: pick.n_games,
  projected_value: pick.projected_value ?? null,
  projected_minutes: pick.projected_minutes ?? null,
  edge: pick.edge ?? null,
});

export const createAgentPickContextFromBetSlip = (
  pick: BetSlipPick,
): AgentPickContext => ({
  player_name: pick.player_name,
  player_team: pick.player_team || "",
  event_id: pick.event_id,
  home_team: pick.home_team,
  away_team: pick.away_team,
  commence_time: pick.commence_time,
  metric: pick.metric,
  threshold: pick.threshold,
  direction: pick.direction,
  probability: pick.probability,
  n_games: pick.n_games,
  projected_value: null,
  projected_minutes: null,
  edge: null,
});
